.github/workflows/on-pull-request-opened-and-synchronize.yml

```diff
@@ -77,35 +79,151 @@ jobs:
         if: ${{ failure() }}
         with:
           name: coverage
-          path: dist/coverage/
+          path: dist/coverage
+
+  fetch-visual-snapshots:
+    name: Fetch Visual Snapshots
+    runs-on: ubuntu-latest
+    steps:
+      # TODO: Remove after this is resolved: https://www.cloudflarestatus.com/incidents/t5nrjmpxc1cj
+      - name: Install AWS CLI
+        run: |
+          curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64-2.22.35.zip" -o "awscliv2.zip"
+          unzip awscliv2.zip
+          sudo ./aws/install --update
+      - name: Configure AWS CLI
+        env:
+          CLOUDFLARE_R2_ACCESS_KEY_ID: ${{ secrets.CLOUDFLARE_R2_ACCESS_KEY_ID }}
+          CLOUDFLARE_R2_SECRET_ACCESS_KEY: ${{ secrets.CLOUDFLARE_R2_SECRET_ACCESS_KEY }}
+        # https://developers.cloudflare.com/r2/examples/aws/aws-cli
+        run: |
+          aws configure set region auto
+          aws configure set output json
+          aws configure set aws_access_key_id $CLOUDFLARE_R2_ACCESS_KEY_ID
+          aws configure set aws_secret_access_key $CLOUDFLARE_R2_SECRET_ACCESS_KEY
+      - run: aws s3 sync s3://${{ vars.CLOUDFLARE_R2_BUCKET_NAME }}/snapshots ./snapshots --endpoint-url ${{ vars.CLOUDFLARE_R2_ENDPOINT }}
+      - name: Upload snapshots
+        uses: actions/upload-artifact@v4
+        with:
+          path: snapshots
+          name: visual-snapshots
+
+  test-visuals:
+    name: Test (Visuals)
+    # Invididual MacOS jobs are likely faster. But MacOS runner concurrency is
+    # limited to 5. So it would be overall slower and we'd soak up all of our
+    # account's MacOS runners.
+    runs-on: ubuntu-latest
+    # Increase sharding if the job takes longer than 5 minutes. We want it to
+    # be no slower than the `test` job, which is our slowest.
+    timeout-minutes: 5
+    needs:
+      - install-dependencies
+      - fetch-visual-snapshots
+    strategy:
+      # We need every job in the matrix to finish so we can produce a full report.
+      fail-fast: false
+      matrix:
+        shardIndex: [1, 2, 3, 4, 5]
+        shardTotal: [5]
```

Comment:
We may need to increase sharding as we add most tests. Guess we'll see. I'll revisit as we go. Maybe we'll get lucky and can decrease this number.